### PR TITLE
Introduce fast mouse polling for lower lag interpolation

### PR DESCRIPTION
**Pull request by mikeday0**
This is my humble attempt to port to Crispy the fine work that @ceski-1 did for Woof. It's not ready yet but I thought it might be prudent to go ahead and get it out there for early review and testing.

EDIT: Calling this one complete. We are still missing vertical mouse and gamepad, but those can come later. The vertical mouse implementation in particular needs some careful study and refactoring across the four games.

This is not as sophisticated as the current implementation in Woof! but I think it stands on its own as a solid improvement to mousefeel for typical KB + M players.

### CODE REVIEW

**Comment by ceski-1** [comment]
Nice starting point, I tested it briefly. Some issues I noticed:
- Mouselook and mousevert feel off (compare capped and uncapped).
- Combining keyboard turning and mouse turning at the same time has visual glitches (occasional "pops").
- The "mouse threshold" concept in Chocolate/Crispy/Woof assumes each update is a tic in duration (1/35th of a second). So it doesn't work correctly with fast polling and the mouse will feel "slow" at higher fps. I sidestep the issue in Woof by ignoring threshold under certain conditions.
- If you follow the mouse input calculations from start to finish, you'll notice that Crispy has rounding error in several stages. First there's `AccelerateMouse()` in `I_ReadMouse()`, then there's `*(mouseSensitivity+5)/10` in `G_Responder()` or `G_FastResponder()`. By the time the error accumulator runs ("carrying" the error), it's too late and ineffective. I prevent this in Woof by doing the calculations as a single step and using doubles as working variables (see `CalcMouseAngle()` for example). The rounding only occurs once, in the carry function.

I'll try to give more feedback later, I know it's WIP right now.

**Comment by mikeday0** [comment]
> * Combining keyboard turning and mouse turning at the same time has visual glitches (occasional "pops").

I can repeat this. It appears to be fixed on my end by moving the `fractionaltic` update back to the end of `I_FinishUpdate()`.

<details>
<summary> Diff </summary>

```diff

diff --git a/src/i_video.c b/src/i_video.c
index ab3c7bc9..b5f9f6f1 100644
--- a/src/i_video.c
+++ b/src/i_video.c
@@ -544,9 +544,6 @@ void I_StartDisplay(void) // [crispy]
 {
     SDL_PumpEvents();
 
-    // [AM] Figure out how far into the current tic we're in as a fixed_t.
-    fractionaltic = I_GetFracRealTime();
-
     if (usemouse && !nomouse && window_focused)
     {
         I_ReadMouseUncapped();
@@ -941,6 +938,9 @@ void I_FinishUpdate (void)
                 }
             }
         }
+        // [AM] Figure out how far into the current tic we're in as a fixed_t.
+        fractionaltic = I_GetFracRealTime();
+
     }
 
     // Restore background and undo the disk indicator, if it was drawn.

```
</details>

Does this fix the "pops" for you?

**Comment by ceski-1** [comment]
> Does this fix the "pops" for you?

That seem to work fine, or like Woof:

```diff
diff --git a/src/i_video.c b/src/i_video.c
index ab3c7bc9..328e6b1e 100644
--- a/src/i_video.c
+++ b/src/i_video.c
@@ -542,11 +542,11 @@ void I_StartTic (void)
 
 void I_StartDisplay(void) // [crispy]
 {
-    SDL_PumpEvents();
-
     // [AM] Figure out how far into the current tic we're in as a fixed_t.
     fractionaltic = I_GetFracRealTime();
 
+    SDL_PumpEvents();
+
     if (usemouse && !nomouse && window_focused)
     {
         I_ReadMouseUncapped();
```

**Comment by JNechaevsky** [comment]
Finally, tried it out on both 60 and 120 Hz monitors. It's a God's job. The difference is notable even on 120 Hz!

File: src/doom/r_main.c [hunk 1]
```diff
@@ -1007,6 +987,23 @@ R_PointInSubsector
...
+{
+  return (
+    // Don't use localview if the player is spying.
+    player == &players[consoleplayer] &&
+    // Don't use localview if the player is dead.
+    player->playerstate != PST_DEAD &&
+    // Don't use localview if the player just teleported.
+    !player->mo->reactiontime &&
+    // Don't use localview if a demo is playing.
+    !demoplayback &&
```
Comment (JNechaevsky) [inline src/doom/r_main.c:1001]: Small thought: probably worth to check for `!demorecording` here as well. At the moment, while demo recording there is a small notable jiterring between frames while camera rotation, most likely brought by "short tics".

Comment (mikeday0) [reply src/doom/r_main.c:1001]: Thanks for the review Julia. I noticed this issue too. I have a fix for the jittering on my local pc that I will clean up and push when I return home this weekend.

**Comment by mikeday0** [comment]
@fabiangreffrath Kindly requesting a review of this PR for inclusion in the next release. Thanks!

**Review by fabiangreffrath — APPROVED** [review]
Wow, this change is massive! I must have somehow missed that it was ready for review, sorry. Looks good to me, thank you. I cannot review every single line change, but I trust you enough to judge the change set just from the looks of it. 😉

**Comment by mikeday0** [comment]
Just noticed this was a non-squashed merge. Should I revert and re-merge with squash?

**Comment by fabiangreffrath** [comment]
> Just noticed this was a non-squashed merge. Should I revert and re-merge with squash?

Na, it's okay. Thanks!

**Pull request merged**